src/providers/trusted_service/context.rs

```diff
@@ -0,0 +1,117 @@
+
+    unsafe impl Send for service_context {}
+    unsafe impl Sync for service_context {}
+}
+
+#[allow(
+    non_snake_case,
+    non_camel_case_types,
+    non_upper_case_globals,
+    clippy::unseparated_literal_suffix,
+    // There is an issue where long double become u128 in extern blocks. Check this issue:
+    // https://github.com/rust-lang/rust-bindgen/issues/1549
+    improper_ctypes,
+    missing_debug_implementations,
+    trivial_casts,
+    clippy::all,
+    unused,
+    unused_qualifications
+)]
+mod ts_protobuf {
+    include!(concat!(env!("OUT_DIR"), "/ts_crypto.rs"));
+}
+
+#[derive(Debug)]
+pub struct Context {
+    rpc_caller: Option<rpc_caller>,
+    service_context: Option<service_context>,
+    rpc_session_handle: Option<MBox<c_void>>,
+}
+
+impl Context {
+    pub fn connect() -> anyhow::Result<Self> {
+        info!("Querying for crypto Trusted Services");
+        let mut status = 0;
+        let service_context = unsafe {
+            service_locator_query(
+                CString::new("sn:tf.org:crypto:0").unwrap().into_raw(),
+                &mut status,
+            )
+        };
+        if service_context == null_mut() || status != 0 {
+            return Err(Error::new(
+                ErrorKind::Other,
+                format!(
+                    "Failed to connect to Trusted Service; status code: {}",
+                    status
+                ),
+            )
+            .into());
+        }
+
+        info!("Starting crypto Trusted Service context");
+        let mut rpc_caller = null_mut();
+        let rpc_session_handle = unsafe { service_context_open(service_context, &mut rpc_caller) };
+        if rpc_caller == null_mut() || rpc_session_handle == null_mut() {
+            return Err(
+                Error::new(ErrorKind::Other, "Failed to start Trusted Service context").into(),
+            );
+        }
+        let rpc_caller_box = unsafe { MBox::from_raw(rpc_caller) };
+        let service_context_box = unsafe { MBox::from_raw(service_context) };
+        let rpc_session_handle = unsafe { MBox::from_raw(rpc_session_handle) };
+        let ctx = Context {
+            rpc_caller: Some(*rpc_caller_box),
+            service_context: Some(*service_context_box),
+            rpc_session_handle: Some(rpc_session_handle),
+        };
+
+        let _ = rpc_caller_box.into_raw();
+        let _ = service_context_box.into_raw();
+        Ok(ctx)
+    }
+}
+
+impl Drop for Context {
+    fn drop(&mut self) {
+        let mut service_context = self.service_context.take().unwrap();
+        let mut rpc_session_handle = self.rpc_session_handle.take().unwrap();
+        let mut _rpc_caller = self.rpc_caller.take().unwrap();
+
+        unsafe { service_context_close(&mut service_context, rpc_session_handle.as_mut_ptr()) };
+
+        unsafe { service_locator_relinquish(&mut service_context) };
+
+        std::mem::forget(service_context);
+        std::mem::forget(_rpc_caller);
```

Comment:
Referring to my previous comment, why are those needed?

Reply:
Same for why we need to remove the `MBox` - those two are essentially structures holding pointers - if they get dropped, you have double frees.

Reply:
Maybe I can put them in some sort of `WeakPointer` structure that doesn't drop them when the structure is dropped

Reply:
Didn't find anything in `std`, will just use `*mut`